src/Npgsql/PostgresException.cs

```diff
@@ -29,7 +27,204 @@ namespace Npgsql
+            AppendLine(nameof(ColumnName), ColumnName);
+            AppendLine(nameof(DataTypeName), DataTypeName);
+            AppendLine(nameof(ConstraintName), ConstraintName);
+            AppendLine(nameof(File), File);
+            AppendLine(nameof(Line), Line);
+            AppendLine(nameof(Routine), Routine);
+
+            return builder.ToString();
+
+            void AppendLine<T>(string propertyName, T propertyValue)
+            {
+                if (!EqualityComparer<T>.Default.Equals(propertyValue, default))
+                    builder.AppendLine().Append("    ").Append(propertyName).Append(": ").Append(propertyValue);
+            }
+        }
+
+        /// <summary>
+        /// Gets a the PostgreSQL error message and code.
+        /// </summary>
+        public override string Message => SqlState + ": " + MessageText;
+
+        /// <summary>
+        /// Specifies whether the exception is considered transient, that is, whether retrying to operation could
+        /// succeed (e.g. a network error). Check <see cref="SqlState"/>.
+        /// </summary>
+        public override bool IsTransient
+        {
+            get
+            {
+                switch (SqlState)
+                {
+                case PostgresErrorCodes.InsufficientResources:
+                case PostgresErrorCodes.DiskFull:
+                case PostgresErrorCodes.OutOfMemory:
+                case PostgresErrorCodes.TooManyConnections:
+                case PostgresErrorCodes.ConfigurationLimitExceeded:
+                case PostgresErrorCodes.CannotConnectNow:
+                case PostgresErrorCodes.SystemError:
+                case PostgresErrorCodes.IoError:
+                case PostgresErrorCodes.SerializationFailure:
+                case PostgresErrorCodes.LockNotAvailable:
+                case PostgresErrorCodes.ObjectInUse:
+                case PostgresErrorCodes.ObjectNotInPrerequisiteState:
+                case PostgresErrorCodes.ConnectionException:
+                case PostgresErrorCodes.ConnectionDoesNotExist:
+                case PostgresErrorCodes.ConnectionFailure:
+                case PostgresErrorCodes.SqlClientUnableToEstablishSqlConnection:
+                case PostgresErrorCodes.SqlServerRejectedEstablishmentOfSqlConnection:
+                case PostgresErrorCodes.TransactionResolutionUnknown:
+                    return true;
+                default:
+                    return false;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the statement which triggered this exception.
+        /// </summary>
+        public NpgsqlStatement Statement { get; internal set; }
+
+        /// <summary>
+        /// Gets a collection of key/value pairs that provide additional PostgreSQL fields about the exception.
+        /// </summary>
+        public override IDictionary Data
+        {
+            get
+            {
+                if (_initializedData == null)
```

Comment:
This is a bit weird to me - any reason to have an `IDictionary` field rather than a simple boolean flag? We could simply update `base.Data` directly as suggested earlier...

Reply:
We could, should we? The flag field occupies the same memory size as a pointer, but with it we do a lot of virtual calls only to access the underlying read only field with lazy initialization. The second reason why I prefer storing a collections is that the `ListDictionaryInternal` class isn't in .NET Standard, so the base property access is just a hack I have to use.

Reply:
I think there aren't any perf considerations here - we're discussing the (lazy) initialization an exception, which is a rare event that isn't meant to be optimized. So to me it makes more sense to write the code in a natural/logical way, rather than attempting to save on virtual calls. This seems like a pretty standard lazy initialization pattern where one would use an `_isInitialized` flag; having an IDictionary field instead gives the impression that it has its own dictionary that's distinct from `base.Data`, which isn't the case...

> The second reason why I prefer storing a collections is that the `ListDictionaryInternal` class isn't in .NET Standard, so the base property access is just a hack I have to use.

I don't understand... Exception already exposes a `Data` property where we can simply set all the fields we want, where does `ListDictionaryInternal` or .NET Standard come in?